### ISSUE TITLE
Cant figure out how to take the JAWS config for DynamoDB and use it in my Serverless project

### ISSUE THREAD

**Issue by contractorwolf** [serverless/serverless#478]
I tried to set the values in my working JAWS project's resources-cf.json file and put them into my Serverless resources-cf.json file but I get this error when I am trying to insert a record:
{
  "errorMessage": "Error in putItem AccessDeniedException: User: arn:aws:sts::237169730438:assumed-role/ServerlessAPI-development-r-IamRoleLambda-14EZJ5RR44SLA/awslambda_923_20160112012315176 is not authorized to perform: dynamodb:PutItem on resource: arn:aws:dynamodb:us-east-1:237169730438:table/users-development"
}

here is the resources-cf.json file from my Serverless project:
```javascript
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "ServerlessAPI resources",
  "Parameters": {
    "ProjectName": {
      "Type": "String",
      "AllowedValues": [
        "ServerlessAPI"
      ],
      "Default": "ServerlessAPI"
    },
    "ProjectDomain": {
      "Type": "String",
      "Default": "signalvehicle.com"
    },
    "Stage": {
      "Type": "String",
      "AllowedValues": [
        "development"
      ]
    },
    "DataModelStage": {
      "Type": "String",
      "AllowedValues": [
        "development"
      ]
    },
    "NotificationEmail": {
      "Type": "String",
      "Default": "removedfromfile@hotmail.com"
    },
    "DynamoRWThroughput": {
      "Type": "String",
      "Default": "1"
    }
  },
  "Metadata": {
    "AWS::CloudFormation::Interface": {
      "ParameterGroups": [
        {
          "Label": {
            "default": "Project Settings"
          },
          "Parameters": [
            "ProjectName",
            "ProjectDomain",
            "Stage",
            "DataModelStage"
          ]
        },
        {
          "Label": {
            "default": "Monitoring"
          },
          "Parameters": [
            "NotificationEmail"
          ]
        },
        {
          "Label": {
            "default": "Database Settings (DynamoDB)"
          },
          "Parameters": [
            "DynamoRWThroughput"
          ]
        }
      ]
    }
  },
  "Resources": {
    "myDynamoDBTable" : {
      "Type" : "AWS::DynamoDB::Table",
      "Properties" : {
        "AttributeDefinitions": [
          {
            "AttributeName": "username",
            "AttributeType": "S"
          }
        ],
        "KeySchema": [
          {
            "AttributeName": "username",
            "KeyType": "HASH"
          }
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": "5",
          "WriteCapacityUnits": "5"
        },
        "TableName": {
          "Fn::Join": [
            "-",
            [
              "users",
              {
                "Ref": "aaDataModelStage"
              }
            ]
          ]
        }
      }
    },
    "IamRoleLambda": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com"
                ]
              },
              "Action": [
                "sts:AssumeRole"
              ]
            }
          ]
        },
        "Path": "/"
      }
    },
    "IamInstanceProfileLambda": {
      "Type": "AWS::IAM::InstanceProfile",
      "Properties": {
        "Path": "/",
        "Roles": [
          {
            "Ref": "IamRoleLambda"
          }
        ]
      }
    },
    "IamGroupLambda": {
      "Type": "AWS::IAM::Group",
      "Properties": {
        "Path": "/"
      }
    },
    "IamPolicyLambda": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyName": {
          "Fn::Join": [
            "_-_",
            [
              {
                "Ref": "Stage"
              },
              {
                "Ref": "ProjectName"
              },
              "lambda"
            ]
          ]
        },
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
              ],
              "Resource": {
                "Fn::Join": [
                  ":",
                  [
                    "arn:aws:logs",
                    {
                      "Ref": "AWS::Region"
                    },
                    "*:*"
                  ]
                ]
              }
            },
            {
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "dynamodb:*Item",
                "dynamodb:Query",
                "dynamodb:Scan"
              ]
            }
          ]
        },
        "Roles": [
          {
            "Ref": "IamRoleLambda"
          }
        ],
        "Groups": [
          {
            "Ref": "IamGroupLambda"
          }
        ]
      }
    }
  },
  "Outputs": {
    "IamRoleArnLambda": {
      "Description": "ARN of the lambda IAM role",
      "Value": {
        "Fn::GetAtt": [
          "IamRoleLambda",
          "Arn"
        ]
      }
    }
  }
}
```

I have a feeling that it may have something to do with this line:
 "Ref": "aaDataModelStage"
(which have corresponding values in my JAWS project, but I have no idea what I might change it to)

I setup my JAWS project according to this:
https://aws.amazon.com/blogs/compute/getting-started-with-jaws-on-amazon-web-services/

Is there a similar explanation on how to setup DynamoDB in Serverless?

Thanks in advance!

**Comment by tmilewski** [comment]
Quick little note that might help you when searching for this information, this isn't unique to JAWS for Serverless. It's actually a template for AWS Cloudformation.

---
**Resources:**

In your `parameters` section, you define `DataModelStage`. That's what should be used instead of `aaDataModelStage`.

Per your definitions of:

```json
"TableName": {
          "Fn::Join": [
            "-",
            [
              "users",
              {
                "Ref": "aaDataModelStage"
              }
            ]
          ]
        }
```

and

```json
"DataModelStage": {
      "Type": "String",
      "AllowedValues": [
        "development"
      ]
    },
```

This will create a DynamoDB table named `users-development`.

---

**Permissioning:**

I'm not 100% on this but `dynamodb:*Item` might not be a thing. You might have to list all permissions out individually. `dynamodb:GetItem`, `dynamodb:PutItem`, etc.

**Comment by c-knowles** [comment]
I think the ```dynamodb:*Item``` may be it as I have something similar working with explicit listings of the various item actions.

If you are interested in restricting the lambda to just that table then that's also also a good idea for security, although it's a bit convoluted in CF:
```
"Resource": {
  "Fn::Join": [
    ":",
    [
      "arn:aws:dynamodb",
      {
        "Ref": "AWS::Region"
      },
      "*",
      {
        "Fn::Join": [
          "",
          [
            "table/",
            {
              "Ref": "myDynamoDBTable"
            },
            "*"
          ]
        ]
      }
    ]
  ]
}
```
You'll also want to define the ```DependsOn``` attribute as part of ```IamPolicyLambda``` and reference the table resource name:
```
"IamPolicyLambda": {
      "Type": "AWS::IAM::Policy",
      "DependsOn": "myDynamoDBTable"
```

**Comment by contractorwolf** [comment]
@tmilewski and @c-knowles thanks for the great advice.  It did turn out to be the "Ref": "aaDataModelStage" (which is how JAWS named these values with the "aa" prefix).  When I switched it to "Ref": "DataModelStage" as described by @tmilewski my insert worked.  I tested it without making changes to the 
```javascript
"Action": [
                "dynamodb:*Item",
                "dynamodb:Query",
                "dynamodb:Scan"
              ]
```
because I had this working correctly in my JAWS project, and it seems to work in Serverless as well.  

@tmilewski I thought that these were just configurations for Serverless and now see that there is more method to the JSON madness.  Here is a link to using the configuration for all the associated projects in case anyone else needs some help:
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/sample-templates-services-us-west-2.html
